### PR TITLE
[#1245] Add new Device Connection API classes

### PR DESCRIPTION
**Pull request by calohmn**
This is for #1245:

Add interfaces and base implementation classes for the new Device Connection API.

### CODE REVIEW

File: client/src/main/java/org/eclipse/hono/client/DeviceConnectionClient.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+    /**
+     * Sets the given gateway as the last gateway that acted on behalf of the given device.
+     * <p>
+     * If a device connects directly instead of through a gateway, the device identifier itself is to be used as value
+     * for the <em>gatewayId</em> parameter.
+     *
+     * @param deviceId The device id.
+     * @param gatewayId The gateway id (or the device id if the last message came from the device directly).
+     * @param context The currently active OpenTracing span. An implementation
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/DeviceConnectionClient.java]: can this be `null`?

File: client/src/main/java/org/eclipse/hono/client/DeviceConnectionClient.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+     *         should use this as the parent for any span it creates for tracing
+     *         the execution of this operation.
+     * @return A future indicating whether the operation succeeded or not.
+     * @throws NullPointerException if tenant id, device id or gateway id is {@code null}.
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/DeviceConnectionClient.java]: there is no *tenant ID* parameter

File: client/src/main/java/org/eclipse/hono/client/DeviceConnectionClient.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+     *         the execution of this operation.
+     * @return A future indicating the result of the operation.
+     *         <p>
+     *         The future will succeed if a response with status 200 has been received from the device connection service.
+     *         In that case the value of the future will contain a <em>gateway-id</em> property with the
+     *         gateway id.
+     *         <p>
+     *         In case a status other then 200 is received, the future will fail with a
+     *         {@link ServiceInvocationException} containing the (error) status code returned by the service.
+     * @throws NullPointerException if tenant id or device id is {@code null}.
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/DeviceConnectionClient.java]: there is no *tenant ID* parameter

File: client/src/main/java/org/eclipse/hono/client/DeviceConnectionClient.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    Future<Void> setLastKnownGatewayForDevice(String deviceId, String gatewayId, SpanContext context);
+
+    /**
+     * Gets the gateway that last acted on behalf of the given device.
+     * <p>
+     * If no last known gateway has been set for the given device yet, a failed future with status <em>Not Found</em>
+     * is returned.
+     *
+     * @param deviceId The device id.
+     * @param context The currently active OpenTracing span. An implementation
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/DeviceConnectionClient.java]: can this be `null`?

File: client/src/main/java/org/eclipse/hono/client/impl/DeviceConnectionClientImpl.java [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+
+    private static final Logger LOG = LoggerFactory.getLogger(DeviceConnectionClientImpl.class);
+
+    /**
+     * Creates a new client for accessing the Device Connection service.
+     * <p>
+     * The client will be ready to use after invoking {@link #createLinks()} or
+     * {@link #createLinks(Handler, Handler)} only.
+     *
+     * @param connection The connection to Hono.
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/DeviceConnectionClientImpl.java]: the connection to the Device Connection service

File: client/src/main/java/org/eclipse/hono/client/impl/DeviceConnectionClientImpl.java [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+     * @param tenantId The identifier of the tenant for which the client should be created.
+     */
+    protected DeviceConnectionClientImpl(final HonoConnection connection, final String tenantId) {
+        super(connection, tenantId);
+    }
+
+    /**
+     * Creates a new client for accessing the Device Connection service.
+     *
+     * @param connection The connection to Hono.
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/DeviceConnectionClientImpl.java]: the connection to the Device Connection service

File: client/src/test/java/org/eclipse/hono/client/impl/DeviceConnectionClientImplTest.java [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+import io.vertx.proton.ProtonHelper;
+import io.vertx.proton.ProtonReceiver;
+import io.vertx.proton.ProtonSender;
+
+
+/**
+ * Tests verifying behavior of {@link DeviceConnectionClientImpl}.
+ *
+ */
+@RunWith(VertxUnitRunner.class)
```
Comment (sophokles73) [inline client/src/test/java/org/eclipse/hono/client/impl/DeviceConnectionClientImplTest.java]: can you please use JUnit5 for new unit tests

File: service-base/src/main/java/org/eclipse/hono/service/deviceconnection/BaseDeviceConnectionService.java [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+
+        Objects.requireNonNull(request);
+
+        switch (DeviceConnectionConstants.DeviceConnectionAction.from(request.getOperation())) {
+        case GET_LAST_GATEWAY:
+            return processGetLastGatewayRequest(request);
+        case SET_LAST_GATEWAY:
+            return processSetLastGatewayRequest(request);
+        default:
+            return processCustomStateMessage(request);
```
Comment (sophokles73) [inline service-base/src/main/java/org/eclipse/hono/service/deviceconnection/BaseDeviceConnectionService.java]: CustomState?
should be CustomCommand or CustomOperation, shouldn't it?

File: client/src/main/java/org/eclipse/hono/client/impl/DeviceConnectionClientImpl.java [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+        final Map<String, Object> properties = createDeviceIdProperties(deviceId);
+        properties.put(MessageHelper.APP_PROPERTY_GATEWAY_ID, gatewayId);
+
+        final Future<DeviceConnectionResult> resultTracker = Future.future();
+        createAndSendRequest(
+                DeviceConnectionConstants.DeviceConnectionAction.SET_LAST_GATEWAY.getSubject(),
+                properties,
+                null,
+                resultTracker,
+                null,
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/DeviceConnectionClientImpl.java]: the intention here is to **never** cache any result, right?

File: client/src/main/java/org/eclipse/hono/client/impl/DeviceConnectionClientImpl.java [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+    public Future<JsonObject> getLastKnownGatewayForDevice(final String deviceId, final SpanContext context) {
+        Objects.requireNonNull(deviceId);
+        final Future<DeviceConnectionResult> resultTracker = Future.future();
+
+        createAndSendRequest(
+                DeviceConnectionConstants.DeviceConnectionAction.GET_LAST_GATEWAY.getSubject(),
+                createDeviceIdProperties(deviceId),
+                null,
+                resultTracker,
+                null,
```
Comment (sophokles73) [inline client/src/main/java/org/eclipse/hono/client/impl/DeviceConnectionClientImpl.java]: again: **never** cache any results?

File: client/src/main/java/org/eclipse/hono/client/impl/DeviceConnectionClientImpl.java (hunk 9 above)
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/impl/DeviceConnectionClientImpl.java]: yes

File: client/src/main/java/org/eclipse/hono/client/impl/DeviceConnectionClientImpl.java (hunk 10 above)
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/impl/DeviceConnectionClientImpl.java]: yes

File: client/src/main/java/org/eclipse/hono/client/impl/DeviceConnectionClientImpl.java (hunk 9 above)
Comment (sophokles73) [reply client/src/main/java/org/eclipse/hono/client/impl/DeviceConnectionClientImpl.java]: then IMHO we should also ignore any `Cache-Directive` header when creating the `DeviceConnectionResult` from the response message and should not support supplying a `CacheProvider` when creating the client, don't you think?

File: client/src/test/java/org/eclipse/hono/client/impl/DeviceConnectionClientImplTest.java (hunk 7 above)
Comment (calohmn) [reply client/src/test/java/org/eclipse/hono/client/impl/DeviceConnectionClientImplTest.java]: Seeing that the similar `RegistrationClientImplTest`, `TenantClientImplTest` and `CredentialsClientImplTest` classes are still not migrated to JUnit5 as well, I would rather migrate all 4 classes together in a subsequent PR.

Comment (sophokles73) [reply client/src/test/java/org/eclipse/hono/client/impl/DeviceConnectionClientImplTest.java]: no need for a PR, just migrate the tests ...

**Comment by calohmn** [comment]
@sophokles73 I've applied the changes according to the above comments (apart from the JUnit5 migration - see above).

File: client/src/main/java/org/eclipse/hono/client/DeviceConnectionClient.java (hunk 1 above)
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/DeviceConnectionClient.java]: Yes, I've adapted the javadoc to point this out.

File: client/src/main/java/org/eclipse/hono/client/DeviceConnectionClient.java (hunk 4 above)
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/DeviceConnectionClient.java]: Yes, I've adapted the javadoc to point this out.

File: client/src/main/java/org/eclipse/hono/client/impl/DeviceConnectionClientImpl.java (hunk 9 above)
Comment (calohmn) [reply client/src/main/java/org/eclipse/hono/client/impl/DeviceConnectionClientImpl.java]: I've amended the commit accordingly.

**Pull request merged**